core/src/main/java/fr/sncf/osrd/envelope/EnvelopePartBuilder.java

```diff
@@ -68,8 +69,15 @@ public void addStep(double position, double speed, double timeDelta) {
     }
 
     @Override
-    public void setEnvelopePartMeta(EnvelopePartMeta meta) {
-        this.meta = meta;
+    public <T extends EnvelopeAttr> void setAttr(Class<T> attrType, T attr) {
+        assert attrType != null : "called setAttr with attrType == null";
+        assert attr != null : "called setAttr with attr == null";
+        attrs.put(attrType, attr);
```

Comment:
```suggestion
    public <T extends EnvelopeAttr> void setAttr(T attr) {
        assert attr != null : "called setAttr with attr == null";
        attrs.put(attr.getClass(), attr);
```

Reply:
If you do this, you can't use inheritance for attribute values anymore. I'd like to avoid that. We would have to ensure all attribute values are final.

Reply:
I'm okay about allowing both APIs.